src/group_theory/solvable.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2021 Jordan Brown, Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jordan Brown, Thomas Browning and Patrick Lutz
+-/
+
+import group_theory.abelianization
+
+/-!
+# Solvable Groups
+
+In this file we introduce the notion of a solvable group. We define a solvable group as one whose
+derived series is eventually trivial. This requires defining the commutator of two subgroups and
+the derived series of a group.
+
+## Main definitions
+
+* `general_commutator H₁ H₂` : the commutator of the subgroups `H₁` and `H₂`
+* `derived_series G n` : the `n`th term in the derived series of `G`, defined by iterating
+  `general_commutator` starting with the top subgroup
+* `is_solvable G` : the group `G` is solvable
+-/
+
+open subgroup
+
+variables {G : Type*} [group G]
+
+section general_commutator
+
+/-- The commutator of two subgroups `H₁` and `H₂`. -/
+def general_commutator (H₁ H₂ : subgroup G) : subgroup G :=
+closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x}
+
+lemma general_commutator_def (H₁ H₂ : subgroup G) :
+  general_commutator H₁ H₂ = closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x} := rfl
+
+instance general_commutator_normal (H₁ H₂ : subgroup G) [h₁ : H₁.normal]
+  [h₂ : H₂.normal] : normal (general_commutator H₁ H₂) :=
+begin
+  let base : set G := {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x},
+  suffices h_base : base = group.conjugates_of_set base,
+  { dsimp only [general_commutator, ←base],
+    rw h_base,
+    exact subgroup.normal_closure_normal },
+  apply set.subset.antisymm group.subset_conjugates_of_set,
+  intros a h,
+  rw group.mem_conjugates_of_set_iff at h,
+  rcases h with ⟨b, ⟨c, hc, e, he, rfl⟩, d, rfl⟩,
+  exact ⟨d * c * d⁻¹, h₁.conj_mem c hc d, d * e * d⁻¹, h₂.conj_mem e he d, by group⟩,
+end
+
```

Comment:
could you please add `general_commutator_comm`?

Reply:
I'm not sure what statement you mean by `general_commutator_comm` or how it relates to this proof.

Reply:
Okay, I'm guessing that by `general_commutator_comm` you meant
```lean
lemma general_commutator_comm (H₁ H₂ : subgroup G) : ⁅H₁, H₂⁆ = ⁅H₂, H₁⁆
```
so I've now included a proof of that. But I'm still not sure how it relates to the lines you commented on.

Reply:
Oh, I just realized that `general_commutator_comm` was one of the theorems that Oliver originally requested, but which I somehow forgot to add initially. Sorry for the confusion. I assume this issue is now taken care of.